Added an extended version and fixed a few bugs on jabra-gnp

gdpcastro opened the pull request:
Type of pull request:

- [ ] New plugin (Please include [new plugin checklist](https://github.com/fwupd/fwupd/wiki/New-plugin-checklist))
- [x] Code fix
- [x] Feature
- [ ] Documentation


Fixing a couple of bugs with the plugin when it comes to child devices, and added the code to support an extension to the firmware update protocol. 
Code is still untested due to some internal constraints, so some byte values may be wrong, but I wanted to get the PR started, and I will be able to test next week already.

hughsie commented on plugins/jabra-gnp/fu-jabra-gnp-device.c:
```diff
@@ -69,6 +75,57 @@ _fu_usb_device_get_interface_for_class(FuUsbDevice *usb_device, guint8 intf_clas
+	for (guint i = 0; i < ifaces->len; i++) {
+		g_autoptr(GPtrArray) endpoints = NULL;
+		FuUsbEndpoint *ep1 = NULL;
+		iface = g_ptr_array_index(ifaces, i);
```
can FuUsbInterface move into the loop?

hughsie commented on plugins/jabra-gnp/fu-jabra-gnp-device.c:
```diff
@@ -393,17 +451,72 @@ fu_jabra_gnp_device_read_version(FuJabraGnpDevice *self, GError **error)
+
+	/* some devices append an extra '.' to the version, which can confuse fwupd's formats, so
+	 * remove it */
+	if (version[strlen(version) - 1] == '.')
```
i think we need to check that version is at least one char long.

hughsie commented on plugins/jabra-gnp/fu-jabra-gnp-device.c:
```diff
@@ -393,17 +451,72 @@ fu_jabra_gnp_device_read_version(FuJabraGnpDevice *self, GError **error)
+				  fu_jabra_gnp_device_tx_cb,
+				  FU_JABRA_GNP_MAX_RETRIES,
+				  FU_JABRA_GNP_RETRY_DELAY,
+				  (gpointer)&tx_data,
```
do you need the (gpointer)?

hughsie commented on plugins/jabra-gnp/fu-jabra-gnp-device.c:
```diff
@@ -897,7 +1043,9 @@ fu_jabra_gnp_device_add_child(FuDevice *device, guint dfu_pid, GError **error)
 	fu_device_add_instance_u16(FU_DEVICE(child), "VID", fu_device_get_vid(FU_DEVICE(self)));
+
 	fu_device_add_instance_u16(FU_DEVICE(child), "PID", dfu_pid);
+
```
not sure about all this extra whitespace...

gdpcastro replied:
Completely agree, was having the hardest time with getting the newest clang-format on my development machine, so after an hour I gave up and ignored that hook :( 
Will remove them on the next iteration hopefully

hughsie commented on plugins/jabra-gnp/fu-jabra-gnp-device.c:
```diff
@@ -997,12 +1146,25 @@ fu_jabra_gnp_device_write_image(FuJabraGnpDevice *self,
+			FU_JABRA_GNP_PRELOAD_COUNT,
+			error))
+			return FALSE;
+	}
+	/* self->fwu_protocol == FU_JABRA_GNP_PROTOCOL_EXTENDED_OTA*/
+	else {
```
can the comment go in the indented section? This looks a bit weird.

hughsie commented on plugins/jabra-gnp/fu-jabra-gnp-device.c:
```diff
@@ -1100,6 +1264,14 @@ fu_jabra_gnp_device_set_quirk_kv(FuDevice *device,
+static gboolean
+fu_jabra_gnp_device_ready(FuDevice *device, GError **error)
+{
+	/* apparently setting up the child device fails without this function being explicitly
+	 * defined */
```
eigh? This smells like a workaround.

hughsie replied:
So, lets dig in a bit. `FuJabraGnpDevice` derives from `FuUsbDevice` which in turn derives from `FuUdevDevice`.

`FuUsbDevice` hooks `device_class->ready` as `fu_usb_device_ready`. In that we're checking for emulated, getting the interfaces for the device and adding fallback icons. Could the issue be that a "child device" isn't really a `FuUsbDevice` (and thus doesn't have interfaces)?

I think the "proper" way to fix this would be to define a new GObject of "`FuJabraGnpChildDevice`" (e.g. adding `fu-jabra-gnp-child-device.[c|h]`) -- I think that's going to make some of the other code less complicated too.

hughsie replied:
We did that.

hughsie commented on plugins/jabra-gnp/fu-jabra-gnp-device.c:
```diff
@@ -578,6 +691,62 @@ fu_jabra_gnp_device_write_crc(FuJabraGnpDevice *self,
+	    .timeout = FU_JABRA_GNP_STANDARD_RECEIVE_TIMEOUT,
+	};
+
+	fu_memwrite_uint32(tx_data.txbuf + 7, crc, G_LITTLE_ENDIAN);
+	fu_memwrite_uint16(tx_data.txbuf + 11, 0x00, G_LITTLE_ENDIAN);
+	fu_memwrite_uint16(tx_data.txbuf + 13, preload_count, G_LITTLE_ENDIAN);
+	fu_memwrite_uint32(tx_data.txbuf + 15, total_chunks, G_LITTLE_ENDIAN);
```
I do wonder if these would be better as a rustgen struct -- e.g. https://fwupd.github.io/libfwupdplugin/tutorial.html#the-rustgen-helper

gdpcastro replied:
I was wondering if all of the plugin should be "converted" to use rustgen, in the same manner we did with our other plugin (jabra-file), was hoping to get your opinion. I could make the changes if we believe it would be better :)

hughsie replied:
I think the rustgen code is very much more maintainable -- and probably the long term direction the project is going in. I'd be fine if you did the rustgen for this one new method and then added the "convert all the others" to the backlog.

gdpcastro commented:
I'm not sure if my comments are not going through as they have a pending note.
But I'm a bit stuck in the suggestion regarding adding a new child device type. Could you maybe help me understand how to get that address properly, when it comes to common functions between the parent and the child device types.
@hughsie

gdpcastro commented:
I understand that the code duplication is way less than ideal, but I couldn't figure out how to share the common part of the code properly, so I could use some advice there.
I will also add the rustgen fix as soon as we sort this out, to save myself some useless copy pasting work

hughsie commented:
I think CI is unhappy :)

gdpcastro commented:
I think the CI failures now are not due to the PR changes, as they seem to be happening in other branches too

superm1 commented:
> I think the CI failures now are not due to the PR changes, as they seem to be happening in other branches too

I don't see issues on main, can you rebase again and check?

gdpcastro commented:
Ah, I see now. I broke the emulation of our device, which would be expected, as we made some additions to the transfer protocol. I assume a new emulation would need to be created based on these changes, so I could look into that when the code looks good, and before merging

gdpcastro commented:
I seem to be unable to create a correct emulation file. I followed the steps described here https://github.com/fwupd/fwupd/blob/main/docs/device-emulation.md and whenever I try to load the created emulation data, I get this error:
`failed to load emulation data: no backend specified FuUsbDevice` 
Any advice on how to get the correct emulation data to be created would be appreciated

gdpcastro commented:
Hi @hughsie and @superm1 
Could I get some help with generating the emulation files for the updated version of the write-firmware procedure? That's the step blocking CI, and that I can't quite understand how to fix.
I have updated the write-firmware code, as it now involves an extra step, and therefore the emulation fails as the steps don't match anymore.
I have tried following the steps to generate an emulation zip here: https://github.com/fwupd/fwupd/blob/main/docs/device-emulation.md (I am on Ubuntu 24.04, using the development environment if it matters)
And the resulting zip file does not seem to have the same structure as the one present in LVFS and being used for the CI test. The files are the same, but upon opening the `.json` files I can see that the structure is rather different, which probably is the reason I get `failed to load emulation data: no backend specified FuUsbDevice`
I have removed the incorrect emulation from LVFS to avoid issues, but I could upload another version again, in case it could help us understand the problem.
For what it's worth, I also tried generating the emulation from main, and also using WireShark captures, but both failed. Following the steps on main creates the same incorrect emulation.zip, while in the WireShark capture method I get stuck on the `pcap2emulation.py`, so my assumption is that there is something wrong with my environment when creating the emulation files. And as a final note, I have tried also building and creating the emulation from outside the development environment, but run into the same issue.

hughsie commented:
> Could I get some help with generating the emulation files for the updated version of the write-firmware procedure

Sure -- I think the best thing would be to rebase against main and then try again -- there were a couple of emualtion bugs fixed recently and I think maybe you're hitting that. I think I have the Evolve 75 device myself somewhere, right?

hughsie commented:
> I think I have the Evolve 75 device myself somewhere, right?

I do, and the emulation seemed to work after rebasing. I'm just fixing up some style issues and will force push with the new emulation. -><- close.

hughsie commented:
> so I will first be able to do that middle of next week

Sounds good to me; the branch I just pushed LGTM now. Thanks.

gdpcastro commented:
Sorry about the delay on this - I unfortunately got sick and didn't have the opportunity to go into the office to run a final test on the correct devices. I plan to do so on Monday :)

gdpcastro commented:
The bug is maybe a bit too elusive right now. It's not a regression in any way as it's the newly added part that has the bug, and it does not reach any devices as of right now, so I think I might merge the code after I run a test or two and then start a new PR for the bug fix soon.

gdpcastro commented:
Hi @hughsie could we get this approved again? Had to fix a tiny error in the implementation that I caught when running some tests

gdpcastro commented:
Hi @hughsie yeah, had a small oopsie on the merge, will fix those now. 
I think I can get the emulation done today. There will be some 10% on the original gnp-device file that will be uncovered until we fix a bug we caught internally, hopefully that's not an issue, and we are working on a fix already

gdpcastro commented:
I unfortunately don't really understand how to create the emulation files correctly, as the code refers to a child device. I tried to tag both the parent and the child for emulation, install the firmware on the child, save the emulation data, but then the load fails. I will add the data to LVFS and the .json anyway, as I will be OoO next week, and maybe the data could be useful if anyone looks at it

gdpcastro commented:
Hi @hughsie 
I've been looking at this since coming back from vacation. I am thinking it's not possible to create an emulation for a child device that exists only as a logical ID, and not an actual FuUsbDevice with a physical ID. Am I understanding so correctly? If so it may not be possible to increase the coverage of the code in the state it is now, as most of the untested code (specifically the one in fu-jabra-gnp-child-device.c) is unreachable without being able to emulate a non FuUsbDevice.
